contrib/platform/src/com/sun/jna/platform/win32/WinGDI.java

```diff
@@ -28,18 +28,97 @@
 import com.sun.jna.Structure;
 import com.sun.jna.Structure.FieldOrder;
 import com.sun.jna.platform.win32.WinNT.HANDLE;
-import com.sun.jna.platform.win32.WinDef.HBITMAP;
-import com.sun.jna.platform.win32.WinDef.RECT;
+import com.sun.jna.Union;
+
+import static com.sun.jna.platform.win32.WinDef.*;
 
 /**
  * Ported from WinGDI.h.
  * Microsoft Windows SDK 6.0A.
  * @author dblock[at]dblock.org
  * @author Andreas "PAX" L&uuml;ck, onkelpax-git[at]yahoo.de
+ * @author SSHTOOLS Limited, support@sshtools.com
  */
 public interface WinGDI {
     int RDH_RECTANGLES = 1;
 
+    @FieldOrder({ "dmDeviceName", "dmSpecVersion", "dmDriverVersion", "dmSize", "dmDriverExtra", "dmFields", "dmUnion1", "dmColor",
+            "dmDuplex", "dmYResolution", "dmTTOption", "dmCollate", "dmFormName", "dmLogPixels", "dmBitsPerPel", "dmPelsWidth",
+            "dmPelsHeight", "dummyunionname2", "dmDisplayFrequency", "dmICMMethod", "dmICMIntent", "dmMediaType", "dmDitherType",
+            "dmReserved1", "dmReserved2", "dmPanningWidth", "dmPanningHeight" })
+
+    public static class DEVMODE extends Structure {
+        public byte[] dmDeviceName = new byte[Winspool.CCHDEVICENAME];
+        public WORD dmSpecVersion;
+        public WORD dmDriverVersion;
+        public WORD dmSize;
+        public WORD dmDriverExtra;
+        public DWORD dmFields;
+        public DUMMYUNIONNAME dmUnion1;
+
+        public static class DUMMYUNIONNAME extends Union {
+            public DUMMYSTRUCTNAME dummystructname;
+
+            @FieldOrder({ "dmOrientation", "dmPaperSize", "dmPaperLength", "dmPaperWidth", "dmScale", "dmCopies", "dmDefaultSource",
+                    "dmPrintQuality" })
+            public static class DUMMYSTRUCTNAME extends Structure {
```

Comment:
Microsoft actually uses `DUMMYSTRUCTNAME` directly in the API docs.  I've noticed several API uses (such as Eclipse's SWT) doesn't use the nested structs, but I couldn't get it working without it, possibly due to the `dmPosition` being defined twice in the struct.

The nested structs seem to work just fine in testing, but happy to refactor if preferred.

Reply:
Unions are fine, but generally you should try to add code in the constructor that detects which union member needs to be read, and then reads it.   That looks a little bit complex here... do you read fields and then based on which bits are set choose which union member applies?

Reply:
Please excuse my lack of experience here.  This code is a direct cherry-pick from [a project called `damage`](https://github.com/sshtools/damage/blob/701199ca45fb5994d778402ec497dd58b732fed2/src/main/java/org/x/ExWinGDI.java).  I had originally begun writing my own, but was having a hard time getting the structs right.

What I do know is that for the printers I've tested this with, the Unions/Structs fill up just fine with data.  What I don't know is what you mean by "read fields and then choose".

I have a small unit-test (not unit-test worthy) that spits these out and I'm not running into any issues.

Here's an example output from `PRINTER_INFO_9` where this applies:

```
[PRINTER_INFO_9]
  Microsoft XPS Document Writer:
    dmDeviceName: x&�h
    78260792680200004d006900630072006f0073006f00660074002000580050005300200044006f00630075006d0065006e007400200057007200690074006500
    dmSpecVersion: 114
    dmDriverVersion: 0
    dmSize: 109
    dmDriverExtra: 0
    dmFields: 100860929
    dmUnion1:  (Union)
      dummystructname:  (Structure)
      dmOrientation: 0
      dmPaperSize: 0
      dmPaperLength: 0
      dmPaperWidth: 0
      dmScale: 0
      dmCopies: 0
      dmDefaultSource: 0
      dmPrintQuality: 0
      dmPosition:  (Structure)
      x: 0
      y: 0
      dummystructname2:  (Structure)
      dmPosition:  (Structure)
      x: 0
      y: 0
      dmDisplayOrientation: 0
      dmDisplayFixedOutput: 0
    dmColor: 100
    dmDuplex: 1
    dmYResolution: 15
    dmTTOption: 600
    dmCollate: 2
    dmFormName: 
    01005802030000004c00650074007400650072000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
    dmLogPixels: 0
    dmBitsPerPel: 0
    dmPelsWidth: 0
    dmPelsHeight: 0
    dummyunionname2:  (Union)
      dmDisplayFlags: 0
      dmNup: 0
    dmDisplayFrequency: 0
    dmICMMethod: 1
    dmICMIntent: 0
    dmMediaType: 1
    dmDitherType: 2
    dmReserved1: 1
    dmReserved2: -1
    dmPanningWidth: 877873479
    dmPanningHeight: 0
```

... and here's the code I'm using to iterate the fields:

```java
/**
     * Crude convenience method for echoing all struct fields to console
     */
    public static void iterateFields(Structure info, int indent) throws Throwable {
        if(info instanceof PRINTER_INFO_9) {
            // PRINTER_INFO_9 stores all data in DEVMODE Struct
            info = ((PRINTER_INFO_9)info).pDevMode;
        } else if(info instanceof PRINTER_INFO_3) {
            // PRINTER_INFO_3 stores all data in SECURITY_DESCRIPTOR_RELATIVE Struct
            info = ((PRINTER_INFO_3)info).pSecurityDescriptor;
        }

        // Iterate all fields
        for(Field f : info.getClass().getFields()) {
            // Skip internal-use fields
            if(f.getName().startsWith("ALIGN_")) {
                continue;
            }

            System.out.print(spaces(indent + 4) + f.getName() + ": ");
            Object o = f.get(info);
            if(f.getName().equals("Status") || f.getName().equals("dwStatus")) {
                int dwStatus = o instanceof WinDef.DWORD ? ((WinDef.DWORD)o).intValue() : (Integer)o;
                System.out.println(o);
                if(dwStatus == 0) {
                    System.out.println(spaces(indent + 6) + "PRINTER_STATUS_NORMAL");
                    continue;
                }
                for(Field w : Winspool.class.getDeclaredFields()) {
                    if(w.getName().startsWith("PRINTER_STATUS_")) {
                        int flag = (Integer)(w.get(null));
                        if((flag & dwStatus) == flag) {
                            System.out.println(spaces(indent + 6) + w.getName());
                        }
                    }
                }
            } else if (o instanceof byte[]) {
                // TODO:  Determine why PRINTER_INFO_9's dmDeviceName is corrupted
                System.out.println(Native.toString((byte[])o));
                System.out.println(spaces(4 + indent) + bytesToHex((byte[])o)); // TODO: Remove
            } else if (o instanceof Structure) {
                System.out.println(" (" + o.getClass().getSuperclass().getSimpleName() + ")");
                iterateFields((Structure)o, 2);
            } else {
                System.out.println(o);
            }
        }
```

Reply:
There's readable UTF16 partway into the output:  `4d006900630072006f0073006f0066007400` is "Microsoft".  (I'm suspecting the rest includes "XPS Document Writer" but haven't translated that far.)  So the question is why are the first 8 bytes different.

Digging deeper, the [docs for `PRINTER_INFO_9`](https://docs.microsoft.com/en-us/windows/win32/printdocs/printer-info-9) state that the single parameter there is a pointer:
> A pointer to a [DEVMODE](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-devmodea) structure

So the mapping for `PRINTER_INFO_9` should be `DEVMODE.ByReference`.  Or you can map a pointer and use a pointer constructor for `DEVMODE`.

Reply:
@dbwiddis thank you kindly for finding this.  I'm quite lost though.  Looking at the Winspool APIs there are other types which also use pointers.  Do all API pointers need to be changed, or is this just an issue with constructing nested structs?

I also don't understand how one would construct a typed-struct using a `ByReference` pointer.  I have no problem writing a getter and setter, but  I'm lost as to how to implement it, as the existing `@FieldOrder` seems a bit like magic to me.  Is there an example class you can mention that you know does something similar?

I tried to change the Structure to a `Structure.ByReference` but then the `super(new Memory((long)size));` constructor fails.

I found [`OaIdl.java`](https://github.com/java-native-access/jna/blob/5ac61611f3c4a74ca3c6ed05c4ffafd7c6070e30/contrib/platform/src/com/sun/jna/platform/win32/OaIdl.java#L105) which seems to use something similar, but it just looks like it implements the ByReference interface.  Is that all I need to do?

My failed attempts:
```
// i assume not like this
public PRINTER_INFO_8(Pointer pDevMode) {
   this.pDevMode = Structure.newInstance(DEVMODE.class, pDevMode);
}

// i assume not like this either
public Pointer pDevMode;
private DEVMODE devMode = Structure.newInstance(DEVMODE.class, pDevMode);
```

Reply:
Ok... I'm getting further... I simply added the typedef... 

```java
public static class DEVMODE extends Structure {
        public static class ByReference extends DEVMODE implements Structure.ByReference {}
...
```

```java
public static class PRINTER_INFO_9 extends Structure {
        public DEVMODE.ByReference pDevMode;
...
```


... and now I simply get back an `M` as a value for `dmDeviceName`.  This seems better, but I'm still not sure if I'm doing it right. :sweat_smile:

Reply:
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/6345473/160678669-2eb01a12-e20d-439a-85cb-7fc90d5ed964.png">

Reply:
Ok, I can grab the value using the following (I'm not entirely sure how WString fixes the null termination):

```
new WString(new String((byte[])o))
```

The result is:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/6345473/160679422-ed285b6e-7615-4fc6-af4d-a75e3cb05ca2.png">

So I think I'm doing something slightly wrong, but my questions are:
1. Is this the correct way to create a ByReference constructor?
2. What's the best way to snag the wide string once I have the bytes?

Reply:
How do you transform from `byte` to `string`? `Native.toString` assumes C convention and in that case a single `NULL` byte terminates the string. If the bytes encode `UTF-16`  or `UCS-2` (i.e. wide-chars on windows), then for ascii every second byte is a NULL byte. If I have a `byte[]` and now the encoding, I would use the string constructor:

```
new String(dmDeviceName, StandardCharsets.UTF_16);
```

or if `win.ascii` is set

```
new String(dmDeviceName, Native.DEFAULT_CHARSET);
```

Reply:
@matthiasblaesing thanks that makes perfect sense, however for some reason this makes the output worse... 

```java
new String(bytes, Boolean.getBoolean("w32.ascii") ? Native.DEFAULT_CHARSET : StandardCharsets.UTF_16)
```

<img width="542" alt="Screen Shot 2022-03-29 at 2 39 31 PM" src="https://user-images.githubusercontent.com/6345473/160682081-f91b9abf-eed4-4a24-94f1-93965158a6f7.png">

Reply:
This will break for UTF-16. Native.toString makes wrong assumptions (a single null byte will terminate the string)

Reply:
Alternately create a helper function and use `Pointer.getWideString()` on the structure.  See this example:

https://github.com/java-native-access/jna/blob/fb76832972e92a215ecdb12ae4f3d6d0c2bae38d/contrib/platform/src/com/sun/jna/platform/win32/Wtsapi32.java#L231-L296

Reply:
I'm tempted to kill `win32.ascii` with fire. It may have been a good idea to make it possible to access native and unicode API on windows. But we have 2022 I consider any use of the non-unicode API dangerous. We know that other parts of the bindings already make that assumption.

Reply:
As @matthiasblaesing says, apparently `Native.toString()` doesn't work.  But `Pointer.getWideString()` does.   The example I linked for `WTSINFO` is good, but note that the field you're retrieving is the first one so you know the field offset is 0.

So this essentially becomes `pDevMode.getPointer().getWideString(... )`.

Reply:
> > Is this the correct way to create a ByReference constructor?
> 
> Yes. :)
> 
> > What's the best way to snag the wide string once I have the bytes?
> 
> Don't use `new String()`. Use `Native.toString()` and pass the character set.
> 
> ```java
> Native.toString(bytes,  Boolean.getBoolean("w32.ascii") ? Native.DEFAULT_CHARSET : StandardCharsets.UTF_16));
> ```

Unfortunately this doesn't work either.

<img width="1083" alt="image" src="https://user-images.githubusercontent.com/6345473/160685607-4958df93-d934-4e0d-9c73-ecb55be1269c.png">


> This will break for UTF-16. Native.toString makes wrong assumptions (a single null byte will terminate the string)

Thanks.  This field should probably terminate if a null byte exists, it would cause havoc in other places as well, but I'm still a bit perplexed as to how to code this.  So far the `WString` constructor is the closest I've gotten, just one extra character.

> Alternately create a helper function and use `Pointer.getWideString()` on the structure. See this example:

Thanks for the example.  I think I'm missing a bit of context though.  I have a byte array, so I'd need to somehow create a pointer from that to use the `getWideString` helper, no?

If I hard-code this into the struct field looping, it absolutely works.

```
System.out.println(info.getPointer().getWideString(0));
```

... but getting the field's offset from within the struct seems a bit backwards when I already have the byte data, no?   I still feel like I'm doing something wrong.

Reply:
> pDevMode.getPointer().getWideString(... ).

Right, which doesn't scale to any non-zero indices 😞

Reply:
By constructing a new Pointer, I can snag the string value.  I do realize this is slightly more wasteful than tracking the pointer in the struct itself, but I find it slightly more practical.

```java
Pointer pBytes = new Memory(bytes.length);
pBytes.write(0, bytes, 0, bytes.length);
String val = Boolean.getBoolean("w32.ascii") ? pBytes.getString(0) : pBytes.getWideString(0);
System.out.println(val);
```

<img width="385" alt="image" src="https://user-images.githubusercontent.com/6345473/160692024-3756b692-0ff3-49f0-aba0-e7f3ec94d496.png">'

Considering all questions are answered, I'll mark this as resolved.  Please feel free to unresolve if needed. 🥂

Reply:
> @matthiasblaesing thanks that makes perfect sense, however for some reason this makes the output worse...
> 
> ```java
> new String(bytes, Boolean.getBoolean("w32.ascii") ? Native.DEFAULT_CHARSET : StandardCharsets.UTF_16)
> ```
> 
> <img alt="Screen Shot 2022-03-29 at 2 39 31 PM" width="542" src="https://user-images.githubusercontent.com/6345473/160682081-f91b9abf-eed4-4a24-94f1-93965158a6f7.png">

Could you please retest this with:

```java
new String(bytes, Boolean.getBoolean("w32.ascii") ? Native.DEFAULT_CHARSET : StandardCharsets.UTF_16BE)
```

and

```java
new String(bytes, Boolean.getBoolean("w32.ascii") ? Native.DEFAULT_CHARSET : StandardCharsets.UTF_16LE)
```

There are two variant how you can map the 16 bit code points of UTF-16 into bytes and either little or big endian encoding is used. I assume that java defaults to one and windows to the other.

Reply:
I can confirm that `UTF_16BE` corrupts the characters identically to `UTF_16`, confirming your suspicion.

I'm leaving this conversation open though since the additional character is appearing with this technique.

Reply:
I can explain it: You are reading uninitialized data (or data you should not read). Here is a reproducer:

```java
public class HexTest {
    public static void main(String[] args) {
	byte[] data = Hex.decode("4d006900630072006f0073006f00660074002000580050005300200044006f00630075006d0065006e007400200057007200690074006500720000006d000000");
	String s = new String(data, StandardCharsets.UTF_16LE);
	s.chars().forEach(i -> System.out.println(i + " / " + ((char) i)));
	System.out.println(s.substring(0, Math.min(s.indexOf("\0"), s.length())));
    }
}
```

The output (only the last lines):

```
101 / e
114 / r
0 / 
109 / m
0 / 
Microsoft XPS Document Writer
```

It can be observerd, that the "m" you are seeing is reported after a terminating "\0" character, so it does not belong to the String. If you split the string at the right place you get the expected value. Contrary to the C definition of a string java can deal with the `NULL` character in strings (If you are interested: https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8).

Reply:
You can see in the hex output the `6d` after a bunch of nulls is the `m`.   The fundamental problem is `new String(byte[])` includes all the bytes and ignores the nulls.   You might be able to `split()` on null and extract the first value, or possibly use `indexOf()` to search for a null and do a substring.  TMTOWTDI.

>  have a byte array, so I'd need to somehow create a pointer from that to use the getWideString helper, no?

The point is that while you have a byte array in your Structure, it's really the native memory you'd be getting.  You actually could, instead of the byte array, put a `new Memory(length)` in your structure and you'd have the pointer right there to call `getWideString()` on.    Or use the fieldoffset method from the `WTSINFO` example.

Reply:
> I can explain it: You are reading uninitialized data (or data you should not read). Here is a reproducer:

In that case, the size I'm asking for and the size I'm getting differ.  Doesn't this suggest that byte[] mapping between native and Java is at fault?  But to take a step back...  How could a data type of unknown size which is only identified by a null terminated character ever possibly know when it's data has been read?

> You might be able to split() on null and extract the first value, or possibly use indexOf() to search for a null and do a substring.

This seems like the right approach, but to that point, isn't this just re-introducing the "problematic" null assumptions?  This really seems like a catch-22.

To that point, it seems other classes use the Pointer's `toWideString()` function just fine, which is what my current code is using:

```java
Pointer pBytes = new Memory(bytes.length);
pBytes.write(0, bytes, 0, bytes.length);
String val = Boolean.getBoolean("w32.ascii") ? pBytes.getString(0) : pBytes.getWideString(0);
System.out.println(val);
```

So correct me if I'm wrong, but any solution which splits on null is going to suffer this described problem:

> This will break for UTF-16. Native.toString makes wrong assumptions (a single null byte will terminate the string)

Unless the point about the original pointer is that it actually **does** have some type of internal reference to the length.  If that's the case, then I would agree, this all belongs on the Struct, but boy that means the struct has to be quite forcibly type-aware, which AFAIR, isn't really how they're normally used.

Reply:
> So correct me if I'm wrong, but any solution which splits on null is going to suffer this described problem:

OK, correcting as asked. :grin:

A null byte is a single byte. That's what you're dealing with in the array when it's actually _two_ null bytes that are significant in UTF-16.

But once you read it into a Java String, the internal representation is now (2 byte) characters.  So you are searching for a null _character_ in Java. 

Repeating again... you have a byte array which represents UTF-16:
 - You can convert the byte array to a (same length) String using the constructor, and parse the first null _character_
 - Or you can read the byte array directly from its underlying memory in the structure using `getWideString()` which properly deals with the null _bytes_.

Reply:
> To put simpler, if dmDeviceName returns anything but it's original series of bytes[], we're misrepresenting the original API, no? (or is this like the DWORD vs int conversation and usability prevails?)

Ok, the writing's on the wall, I'll fix it in the struct.  I realize now that order is guaranteed, so hard-coding the index really belongs there.  😅 .  Is the only way to use a getter, since when I try to add more fields, it crashes.

Reply:
Proposed solution:

```java
        public String getDmDeviceName() {
            int offset = fieldOffset("dmDeviceName");
            return CHAR_WIDTH == 1 ? getPointer().getString(offset) : getPointer().getWideString(offset);
        }

        public String getDmFormName() {
            int offset = fieldOffset("dmFormName");
            return CHAR_WIDTH == 1 ? getPointer().getString(offset) : getPointer().getWideString(offset);
        }
```

Reply:
Resolved via https://github.com/java-native-access/jna/pull/1429/commits/820eb7faa4cccd47c16ded602d7a4ed74a528f6a.

Reply:
> Ok, so from what I understand, this statement:
> 
> > This will break for UTF-16. Native.toString makes wrong assumptions (a single null byte will terminate the string)
> 
> That `Native.toString()` aggressively (and incorrectly) stops on a single null byte, even when using UTF-16, but that the assumption that a UTF-16 null character (in this field, specifically) is still never allowed. This makes a lot more sense.

Yes. For `Native.toString(byte[])`. There is a `Native.toString(char[])` that works as expected for Windows wide strings.

> Right, but this still puts the burden on the struct, but moreover makes anyone using the struct's fields to use alternatives, no? 

Remember that there are actually _two_ structures, `DEVMODEW` (for which you have an array of `WCHAR` --> Java `char`) and `DEVMODEA` (for which you have an array of `CHAR` --> Java `byte`).  In some newer functions, [Windows has given up on the -A variant](https://docs.microsoft.com/en-us/windows/win32/intl/unicode-in-the-windows-api?msclkid=08604e4fafb211ec80333ee5c80f6d90).  

> Is it normal to mutate from within the struct? 

That's just something we have done historically to simplify it for consumers of JNA's structure so they don't have to figure out which one to use.  You can either map one structure and do some internal magic like we're doing here, or map both -A and -W, and use a Util function to choose which one to use.  You can see examples of both types in JNA mappings, just search the code base for `w32.ascii`.  The ones in Util classes often choose between -A and -W variants.  The ones in the DLL mappings tend to manipulate within the structure itself.  There are some that are incorrectly just using the -W variant without the suffix, and they don't break because nobody really uses -A anymore, so we're fine with that.

This is all based on an assumption that -A is ASCII.  [But new functions may be using -A for UTF-8](https://docs.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page).  So in the future the `w32.ascii` might not be the right choice and we'll have to figure out if we're in UTF8 or not.  Or just ignore UTF8 support and only map the W versions and be done with it.

> From my understanding, the purpose of a struct was to allow structured data to be consumed in a predictable form, so mutating it under its original name seems misleading to developers wanting to use it.
> 
> To put simpler, if `dmDeviceName` returns anything but it's original series of `bytes[]`, we're misrepresenting the original API, no? (or is this like the `DWORD` vs `int` conversation and usability prevails?)

It is a predictable form. A fixed length CHAR (byte) array for the -A structure, with text terminated by a null byte, and a fixed length WCHAR (char) array for the -W structure, with text terminated by a null WCHAR.

The struggle you are experiencing is in using a `byte[]` of double the size to represent what you really want as a `char[]` with the convenience of a single structure.   If you just map the -W version and use `char[]` it's so much simpler.

Reply:
> Is the only way to use a getter? since when I try to add more fields, it crashes.

I'm not sure what you're asking here.  Ultimately the user has the same access to the `byte[]` field as you do and can do the same manipulations as your getter.  Creating the convenience getter makes it easier for them so they don't spend the same hours you have today trying to figure out why they have an extra `m`. :)

Reply:
> I'm not sure what you're asking here. 

Well, as a struct, it feels odd to add a getter.  It would be nice to be able to have another field that stores it as a Java string as it would be a more intuitive autocomplete, but adding a static field crashes it. 

I understand there's something about a struct must be the size intended, but if we're setting field order manually, it seems like the Structure would be able to accommodate the fields it needs without throwing a size exception.

> If you just map the -W version and use char[] it's so much simpler.

Should I change it to use `char[]` or are you just minimizing how common it is?

I've had fantastic success with JNA up until this point, and I've even dealt with Windows' strings before.  Since JNA is so friendly about automatically converting certain types, would it make sense to allow Structures to have Strings in place of `byte[]` and handle the logic internally?

Reply:
> Well, as a struct, it feels odd to add a getter. It would be nice to be able to have another field that stores it as a Java string as it would be a more intuitive autocomplete, but adding a static field crashes it.

There are, in theory, ways you could add a duplicate String field, override the field order methods to remove that, and overriding `read()` to set the new value, but that seems a lot of work to avoid a getter.  This way reallly matches the API the best. 

> Should I change it to use char[] or are you just minimizing how common it is?

Earlier in this thread we made the comment about getting rid of the -A mappings and special treatment of w32.ascii. I don't think any version since Windows XP really supports it.  So yes, it is that common.  But it would break compatibility to change it all now.  Up to you if you want to do it more simply, but having a switch and a convenience getter isn't really that much extra work.

> would it make sense to allow Structures to have Strings in place of byte[] and handle the logic internally?

String immutability makes that a challenge.  

Also, Strings in structures are already used as pointers to buffers elsewhere in memory, and work seamlessly reading 1-byte or 2-bytes where that is the mapping.  The problem is when the API specifies fixed length array of characters and you try to combine the two structure mappings into one.